byzcoin/contracts/deferred.go

```diff
@@ -0,0 +1,379 @@
+package contracts
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"strconv"
+
+	"go.dedis.ch/cothority/v3/byzcoin"
+	"go.dedis.ch/cothority/v3/darc"
+	"go.dedis.ch/protobuf"
+)
+
+// The deferred contract allows a group of signers to agree on and sign a
+// proposed transaction, the "proposed transaction".
+
+// ContractDeferredID denotes a contract that can aggregate signatures for a
+// "proposed" transaction
+var ContractDeferredID = "deferred"
+
+const defaultNumExecution uint64 = 1
+
+// DeferredData contains the specific data of a deferred contract
+type DeferredData struct {
+	// The transaction that signers must sign and can be executed with an
+	// "executeProposedTx".
+	ProposedTransaction byzcoin.ClientTransaction
+	// If the current block index is greater than this value, any Invoke on the
+	// deferred contract is rejected. This provides an expiration mechanism.
+	ExpireBlockIndex uint64
```

Comment:
the problem with a wall-clock time limit is that the consensus result completely relies on network delay and clock synchronisation. a "late" message will cause honest nodes to reject what they would usually accept, causing consensus to fail. using block index won't have this problem but it's true that if there are no transactions, then no blocks are created. i'm not sure what's the best solution here...

Reply:
I guess we need to rethink what is behind the expiration, what exactly is the purpose ? Is it like a feature for people to use or is it a cleaning/security feature ?

Reply:
i think it's a feature to make sure every deferred contract reaches some "done" state. it's like saying everyone has to sign before the expiry time otherwise this instance become invalid. i don't know how useful it is or the consequences if we don't have this feature.

another way to do something similar is to allow users to remove their proofs. then we don't need to worry about timing 

maybe @nkcr has more ideas?

Reply:
I was writing my case to keep the block index but changed my mind..

As @kc1212 said, this is to ensure that a proposed transaction is eventually done and can not stay hidden unfinished and then cause trouble later. This is not meant as a feature that client can use. For this reason, I would rather specify a timestamp that should provide a roughly expire date. Even if a majority of Conodes have like a day of difference (not likely to happen I think), that shouldn't be problem if my timestamps is in weeks, or months, or more.

Reply:
So, when spawning the contract, specify an absolute `uint64 timestamp`. Then each Conode check the expiration by checking if the provided timestamps is lower than its current one.